apps/settings/js/panels/call/panel.js

```diff
@@ -0,0 +1,472 @@
+  var DialogService = require('modules/dialog_service');
+
+  return function ctor_call_panel() {
+    return SettingsPanel({
+      onInit: function(panel, options) {
+        this._cardIndex = options.cardIndex || 0;
+        this._conns = window.navigator.mozMobileConnections;
+        this._conn = this._conns[this._cardIndex];
+
+        var $ = function(selector) {
+          return panel.querySelector(selector);
+        };
+
+        this._elements = {
+          panel: panel,
+          header: $('gaia-header'),
+          rilCallerId: $('#ril-callerId'),
+          fdnDesc: $('#fdnSettings-desc'),
+          alertPanel: $('#call .cw-alert'),
+          alertPanelSetButton: $('#call .cw-alert .cw-alert-set'),
+          alertPanelConfirmInput:
+            $('#call .cw-alert .cw-alert-checkbox-label input'),
+          alertPanelCancelButton: $('#call .cw-alert .cw-alert-cancel'),
+          alertLabel: $('#menuItem-callWaiting .alert-label'),
+          fdnItem: $('#menuItem-callFdn'),
+          callWaitingItem: $('#menuItem-callWaiting'),
+          callWaitingInput: $('#menuItem-callWaiting .checkbox-label input'),
+          callForwardingItem: $('#menuItem-callForwarding'),
+          callForwardingAnchor: $('#menuItem-callForwarding a'),
+          callBarringItem: $('#menuItem-callBarring'),
+          callerIdItem: $('#menuItem-callerId'),
+          voiceMailItem: $('.menuItem-voicemail'),
+          voiceMailDesc: $('#voiceMail-desc'),
+          voicePrivacyItem: $('#menuItem-voicePrivacyMode'),
+          voicePrivacyInput: $('#menuItem-voicePrivacyMode input')
+        };
+
+        // Set the navigation correctly when on a multi ICC card device.
+        if (DsdsSettings.getNumberOfIccSlots() > 1) {
+          this._elements.header.setAttribute('data-href', '#call-iccs');
+        }
+
+        this._bindPanelReadyEvent();
+        this._bindVisibilityChangeEvent();
+        this._bindCallForwardingMenuItemClickEvent();
+
+        this._addVoiceTypeChangeListeners();
+
+        this._initCallerId();
+        this._initCallWaiting();
+        this._initVoiceMailSettings();
+        this._initVoicePrivacyModeChangeEvent();
+        this._initVoicemailClickEvent();
+        this._initCallWaitingAlertLabelClickEvent();
+        this._initCallWaitingAlertCancelButtonClickEvent();
+
+        CallSettingsTaskScheduler.observe('isLocked', (isLocked) => {
+          this._enableTapOn('callerIdItem', !isLocked);
+          this._enableTapOn('callWaitingItem', !isLocked);
+        });
+
+        this._refreshCallSettingItems();
+      },
+
+      onBeforeShow: function(panel, options) {
+        if (options.cardIndex) {
+          this._cardIndex = options.cardIndex;
+          this._conn = this._conns[options.cardIndex];
+        }
+
+        this._initFDNItem();
+        this._updateNetworkTypeLimitedItemsVisibility(
+          this._conn.voice && this._conn.voice.type);
+      },
+
+      _bindPanelReadyEvent: function() {
```

Comment:
I am thinking to get rid of listening to the `panelready` event and not making the rendering depending on the previous page.

I believe after we moved the time consuming operations (call forwarding) to a separate panel, querying the remaining call waiting and caller id requires much less time than before. Maybe we can simply call `_refreshCallSettingItems` in `onBeforeShow`, WDYT?